examples/vision/bit.py

```diff
@@ -0,0 +1,255 @@
+from 1 example per class to 1M total examples.
+"""
+
+"""
+## Setup
+"""
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+
+import tensorflow as tf
+from tensorflow import keras
+import tensorflow_hub as hub
+import tensorflow_datasets as tfds
+
+tfds.disable_progress_bar()
+
+SEEDS = 42
+
+np.random.seed(SEEDS)
+tf.random.set_seed(SEEDS)
+
+"""
+## Gather Flower Dataset
+"""
+
+train_ds, validation_ds = tfds.load(
+    "tf_flowers", split=["train[:85%]", "train[85%:]"], as_supervised=True,
+)
+
+"""
+## Visualise the dataset
+"""
+
+plt.figure(figsize=(10, 10))
+for i, (image, label) in enumerate(train_ds.take(9)):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(image)
+    plt.title(int(label))
+    plt.axis("off")
+
+"""
+## Define hyperparameters
+"""
+
+RESIZE_TO = 384
+CROP_TO = 224
+BATCH_SIZE = 64
+STEPS_PER_EPOCH = 10
+AUTO = tf.data.AUTOTUNE  # optimise the pipeline performance
+NUM_CLASSES = 5  # number of classes
+SCHEDULE_LENGTH = (
+    500  # we will train on lower resolution images and will still attain good results
+)
+SCHEDULE_BOUNDARIES = [
+    200,
+    300,
+    400,
+]  # more the dataset size the schedule length increase
+
+"""
+## Preprocessing helper functions
+"""
+
+SCHEDULE_LENGTH = SCHEDULE_LENGTH * 512 / BATCH_SIZE
+
+
+@tf.function
+def preprocess_train(image, label):
+    image = tf.image.random_flip_left_right(image)
+    image = tf.image.resize(image, (RESIZE_TO, RESIZE_TO))
+    image = tf.image.random_crop(image, (CROP_TO, CROP_TO, 3))
+    image = image / 255.0
+    return (image, label)
+
+
+@tf.function
+def preprocess_test(image, label):
+    image = tf.image.resize(image, (RESIZE_TO, RESIZE_TO))
+    image = image / 255.0
+    return (image, label)
+
+
+DATASET_NUM_TRAIN_EXAMPLES = tf.data.experimental.cardinality(train_ds).numpy()
+
+repeat_count = int(
+    SCHEDULE_LENGTH * BATCH_SIZE / DATASET_NUM_TRAIN_EXAMPLES * STEPS_PER_EPOCH
+)
+repeat_count += 50 + 1  # To ensure at least there are 50 epochs of training
+
+"""
+## Create data pipeline for training
+"""
+
+# Training Pipeline
+pipeline_train = (
+    train_ds.shuffle(10000)
+    .repeat(repeat_count)  # Repeat dataset_size / num_steps
+    .map(preprocess_train, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+# Validation Pipeline
+pipeline_validation = (
+    validation_ds.map(preprocess_test, num_parallel_calls=AUTO)
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+"""
+## Visualise the training samples
+"""
+
+image_batch, label_batch = next(iter(pipeline_train))
+
+plt.figure(figsize=(10, 10))
+for n in range(25):
+    ax = plt.subplot(5, 5, n + 1)
+    plt.imshow(image_batch[n])
+    plt.title(label_batch[n].numpy())
+    plt.axis("off")
+
+"""
+## Load model into KerasLayer
+"""
+
+bit_model_url = "https://tfhub.dev/google/bit/m-r50x1/1"
+bit_module = hub.KerasLayer(bit_model_url, trainable=True)
```

Comment:
One thing I don't understand: the model is trainable, and you add a randomly initialized layer on top.

Normally, the process for training such models is to first set the base model as non-trainable, then add the classifier on top, train it, then unfreeze (all or part) of the base, and train again.

If you don't do this, the pretrained weights will be invalidated by very large gradient update at the beginning of the training.